### PR TITLE
fix: ignore calls with age older than lifetime

### PR DESCRIPTION
**Pull request by td-famedly**
Fixes: https://github.com/famedly/product-management/issues/1469

### CODE REVIEW

File: lib/src/client.dart [hunk 1]
```diff
@@ -2118,6 +2118,17 @@ class Client extends MatrixApi {
                     'Ommit "${event.type}" event for an already terminated call');
                 return true;
               }
+              final age = callEvent.unsigned?.tryGet<int>('age') ??
```
Comment (nico-famedly) [inline lib/src/client.dart:2121]: Why are we not using the origin_server_ts?

Comment (nico-famedly) [reply lib/src/client.dart:2121]: I guess the clocks between us and the server could be wrong?

Comment (td-famedly) [reply lib/src/client.dart:2121]: > I guess the clocks between us and the server could be wrong?

Only reason I could come up with and the docs mention the same

File: lib/src/voip/call.dart [hunk 2]
```diff
@@ -32,16 +32,16 @@ const String voipProtoVersion = '1';
 
 class Timeouts {
   /// The default life time for call events, in millisecond.
-  static const lifetimeMs = 10 * 1000;
+  static const lifetimeMs = Duration(milliseconds: 10000);
 
   /// The length of time a call can be ringing for.
-  static const callTimeoutSec = 60;
+  static const callTimeoutMs = Duration(milliseconds: 60000);
```
Comment (nico-famedly) [inline lib/src/voip/call.dart:38]:
```suggestion
  static const callTimeout = Duration(seconds: 60);
```

File: lib/src/voip/call.dart [hunk 3]
```diff
@@ -32,16 +32,16 @@ const String voipProtoVersion = '1';
 
 class Timeouts {
   /// The default life time for call events, in millisecond.
-  static const lifetimeMs = 10 * 1000;
+  static const lifetimeMs = Duration(milliseconds: 10000);
```
Comment (nico-famedly) [inline lib/src/voip/call.dart:35]:
```suggestion
  static const lifetime = Duration(seconds: 10);
```

File: lib/src/voip/call.dart [hunk 4]
```diff
@@ -32,16 +32,16 @@ const String voipProtoVersion = '1';
...
 
   /// The length of time a call can be ringing for.
-  static const callTimeoutSec = 60;
+  static const callTimeoutMs = Duration(milliseconds: 60000);
 
   /// The delay for ice gathering.
-  static const iceGatheringDelayMs = 200;
+  static const iceGatheringDelayMs = Duration(milliseconds: 200);
```
Comment (nico-famedly) [inline lib/src/voip/call.dart:41]:
```suggestion
  static const iceGatheringDelay = Duration(milliseconds: 200);
```

File: lib/src/voip/call.dart [hunk 5]
```diff
@@ -32,16 +32,16 @@ const String voipProtoVersion = '1';
...
 
   /// Delay before createOffer.
-  static const delayBeforeOfferMs = 100;
+  static const delayBeforeOfferMs = Duration(milliseconds: 100);
```
Comment (nico-famedly) [inline lib/src/voip/call.dart:44]:
```suggestion
  static const delayBeforeOffer = Duration(milliseconds: 100);
```

**Review by nico-famedly — APPROVED** [review]
Some nits and maybe you can add a test for it although it might be possible it is already covered, since you modified some tests?

File: lib/src/voip/voip.dart [hunk 6]
```diff
@@ -263,16 +267,22 @@ class VoIP {
...
     // by terminate.
     currentCID = callId;
-
-    await newCall.initWithInvite(
-        callType, offer, sdpStreamMetadata, lifetime, confId != null);
+    try {
+      await newCall.initWithInvite(
+          callType, offer, sdpStreamMetadata, lifetime, confId != null);
+    } catch (e, s) {
+      Logs().e('[VOIP] initWithInvite failed', e, s);
```
Comment (nico-famedly) [inline lib/src/voip/voip.dart:274]: So the call init failed, why do we show a call page? Shouldn't we error out? :D

File: test/calls_test.dart [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+                  'offer': {'type': 'offer', 'sdp': 'sdp'}
+                },
+                senderId: '@alice:testing.com',
+                eventId: 'newevent',
+                originServerTs: DateTime.utc(1969),
+              )
+            ]))
+          })));
+      await Future.delayed(Duration(seconds: 3));
+      expect(voip.currentCID, null);
```
Comment (nico-famedly) [inline test/calls_test.dart:81]: Is there no way to synchronize this?

Comment (td-famedly) [reply test/calls_test.dart]: added a comment as discussed

**Pull request merged**